metric_learn/rca.py

```diff
@@ -8,47 +8,68 @@
 subsets of points that are known to belong to the same class.
 
 'Learning distance functions using equivalence relations', ICML 2003
+'Learning a Mahalanobis metric from equivalence constraints', JMLR 2005
 """
 
 from __future__ import absolute_import
 import numpy as np
 from six.moves import xrange
+from sklearn import decomposition
+import warnings
 
 from .base_metric import BaseMetricLearner
 from .constraints import Constraints
 
 
+# mean center each chunklet separately
+def _chunk_mean_centering(data, chunks, num_chunks):
+  chunk_mask = chunks != -1
+  chunk_data = data[chunk_mask]
+  chunk_labels = chunks[chunk_mask]
+  for c in xrange(num_chunks):
+    mask = chunk_labels == c
+    chunk_data[mask] -= chunk_data[mask].mean(axis=0)
+
+  return chunk_mask, chunk_data
+
+
 class RCA(BaseMetricLearner):
   """Relevant Components Analysis (RCA)"""
-  def __init__(self, dim=None):
+  def __init__(self, dim=None, pca_comps=None):
     """Initialize the learner.
 
     Parameters
     ----------
     dim : int, optional
         embedding dimension (default: original dimension of data)
+    pca_comps : int, float, None or string
+        see the sklearn documentation
     """
-    self.params = {
-      'dim': dim,
-    }
+    self.params = {'dim': dim, 'pca_comps': pca_comps}
 
   def transformer(self):
     return self._transformer
 
-  def _process_inputs(self, X, Y):
-    X = np.asanyarray(X)
-    self.X = X
-    n, d = X.shape
+  def _process_data(self, data):
+    data = np.asanyarray(data)
+    self.X = data
+    n, d = data.shape
+    return data, d
 
+  def _process_chunks(self, data, chunks):
+    chunks = np.asanyarray(chunks)
+    num_chunks = chunks.max() + 1
+    chunk_mask, chunk_data = _chunk_mean_centering(data, chunks, num_chunks)
+    return chunk_mask, chunk_data
+
+  def _process_parameters(self, d):
     if self.params['dim'] is None:
       self.params['dim'] = d
-    elif not 0 < self.params['dim'] <= d:
-      raise ValueError('Invalid embedding dimension, must be in [1,%d]' % d)
-
-    Y = np.asanyarray(Y)
-    num_chunks = Y.max() + 1
-
-    return X, Y, num_chunks, d
+    if self.params['dim'] > d:
+      self.params['dim'] = d
+      warnings.warn('dim must be smaller than the data dimension. ' +
+                    'dim is set to %d.' % (d))
+    return self.params['dim']
```

Comment:
We should also check that `self.params['dim'] >= 1`, and raise a ValueError if it's not.